src/OrchardCore/OrchardCore.DisplayManagement.Liquid/Tags/ShapePagerTag.cs

```diff
@@ -1,75 +1,72 @@
+        public static async ValueTask<Completion> WriteToAsync(ValueTuple<Expression, List<FilterArgument>> arguments, TextWriter writer, TextEncoder encoder, TemplateContext context)
         {
-            var objectValue = (await expression.EvaluateAsync(context)).ToObjectValue() as dynamic;
+            var objectValue = (await arguments.Item1.EvaluateAsync(context)).ToObjectValue() as dynamic;
 
             if (objectValue is Shape shape)
             {
-                var arguments = (FilterArguments)(await new ArgumentsExpression(args).EvaluateAsync(context)).ToObjectValue();
-
                 if (shape.Metadata.Type == "PagerSlim")
                 {
-                    foreach (var name in arguments.Names)
+                    foreach (var argument in arguments.Item2)
                     {
-                        var argument = arguments[name];
-                        var propertyName = name.ToPascalCaseUnderscore();
+                        var propertyName = argument.Name.ToPascalCaseUnderscore();
 
                         if (_properties.Contains(propertyName))
                         {
-                            objectValue[propertyName] = argument.ToStringValue();
+                            objectValue[propertyName] = (await argument.Expression.EvaluateAsync(context)).ToStringValue();
                         }
                     }
                 }
 
+                var expressions = new NamedExpressionList(arguments.Item2);
+
                 if (shape.Metadata.Type == "PagerSlim" || shape.Metadata.Type == "Pager")
                 {
-                    if (arguments.Names.Contains("item_classes"))
+                    if (expressions.HasNamed("item_classes"))
                     {
-                        var classes = arguments["item_classes"];
+                        var itemClasses = await expressions["item_classes"].EvaluateAsync(context);
 
-                        if (classes.Type == FluidValues.String)
+                        if (itemClasses.Type == FluidValues.String)
                         {
-                            var values = classes.ToStringValue().Split(' ', StringSplitOptions.RemoveEmptyEntries);
+                            var values = itemClasses.ToStringValue().Split(' ', StringSplitOptions.RemoveEmptyEntries);
 
                             foreach (var value in values)
                             {
                                 objectValue.ItemClasses.Add(value);
                             }
                         }
-                        else if (classes.Type == FluidValues.Array)
+                        else if (itemClasses.Type == FluidValues.Array)
                         {
-                            foreach (var value in classes.Enumerate())
+                            foreach (var value in itemClasses.Enumerate())
```

Comment:
Might make sense to open up some of the helpers I introduced as internal, like `ToStringList()` or similar, could use the return value automatically and get correct size list to use